rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+reduce op.
+
+### Implementation details
+
+From the implementation POV of the proposed spec, we note that
+`input_conversion` and `output_conversion` can very well be optional with
+default values as identity functions. For example, the following code snippet
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+    ^reduce_computation(
+            %lhs: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
+            %rhs: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
+        %add = "stablehlo.add"(%lhs, %rhs)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
+               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
+            -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
+    }) {
+        dimensions = dense<...> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>
+```
+
+should be interpreted as
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+    ^input_conversion(
+            %input: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
+        "stablehlo.return"(%input)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>) -> ()
+
+    }, {
+    ^reduce_computation(
+            %lhs: tensor<!quant.uniform<ui8:f32, input_scale:0>>,
+            %rhs: tensor<!quant.uniform<ui8:f32, input_scale:0>>):
+        %add = "stablehlo.add"(%lhs, %rhs)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>,
+               tensor<!quant.uniform<ui8:f32, input_scale:0>>)
+            -> tensor<!quant.uniform<ui8:f32, input_scale:0>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>) -> ()
+    }, {
+    ^output_conversion(
+            %intermediate_result: tensor<!quant.uniform<ui8:f32, input_scale:0>>):
+        "stablehlo.return"(%intermediate_result)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
+    }) {
+        dimensions = dense<...> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>
+```
+
+Note that with default values, the  input/result type of `reduce` op matches
+with the argument or the result type of the `reduce_computation`, including the
+quantization parameters.
+
+It is important to enforce the following aspect in the implementations: both
+conversion functions must be provided or both must be missing. If this is not
+followed, and there are two functions provided in the IR, it would be difficult
+to determine which function is intended for which purpose.
```

Comment:
I don't understand this paragraph.  From the use case POV, it is reasonable to have on-the-fly input conversion but leave output in accumulated type.

If you are talking about MLIR ODS, we can use `VariadicRegion` and distinguish them with the order and signature.  The regions can be either `[input: (S)->T, reduce: (T, T)->T, output: (T)->V]`, `[reduce: (T, T)->T, output: (T)->V]`, `[input: (S->T), reduce: (T, T)->T]`, or `[reduce: (T, T)->T]`

Reply:
In all three are present then yes the order is the way to go. 

```
[reduce: (T, T)->T, output: (T)->V], 
[input: (S->T), reduce: (T, T)->T]
```
How to distinguish between the above two just based on types if `T = V  = S`? Also we cannot confuse one with the other as these  (input/output/reduce) functions can have different semantics.

My proposal here is   size(VariadicRegion) = 1 or 3. Let me know if I am missing anything.

Reply:
reduce is binary, other are unary.

Reply:
That is right. Actually my confusion was how to deduce if a unary function is a input or output conversion function if only two functions are provided. 
I understand that relative order of a those function w.r.t the reduce block is the key here, given that reduce block is mandatory. Having said that I agree that we do not need this restriction. I will update.